File: android-ui/choose-currency/src/androidMain/kotlin/io/github/viacheslav/chugunov/moniqo/ui/choosecurrency/screen/ChooseCurrencyMapper.kt
package io.github.viacheslav.chugunov.moniqo.ui.choosecurrency.screen

import io.github.viacheslav.chugunov.moniqo.core.model.Currency
import io.github.viacheslav.chugunov.moniqo.core.model.CurrencyInfo
import io.github.viacheslav.chugunov.moniqo.core.model.CurrencyRates
import io.github.viacheslav.chugunov.moniqo.ui.core.StringProvider
import io.github.viacheslav.chugunov.moniqo.ui.core.model.CurrencyMeta

internal interface ChooseCurrencyMapper {
    fun toContent(
        rates: CurrencyRates,
        recentCodes: List<String>,
    ): ChooseCurrencyState.Content
}

internal class ChooseCurrencyMapperImpl(
    private val stringProvider: StringProvider,
) : ChooseCurrencyMapper {
    override fun toContent(
        rates: CurrencyRates,
        recentCodes: List<String>,
    ): ChooseCurrencyState.Content {
        val allCurrencies =
            buildList {
                add(toCurrencyInfo(rates.baseCurrency))
                rates.rates.forEach { add(toCurrencyInfo(it.currency)) }
            }.sortedBy { it.code }

        val currencyByCode = allCurrencies.associateBy { it.code }
        val recent = recentCodes.mapNotNull { currencyByCode[it] }

        return ChooseCurrencyState.Content(
            recentCurrencies = recent,
            currencies = allCurrencies,
        )
    }

    private fun toCurrencyInfo(currency: Currency): CurrencyInfo {
        val code = currency.name.uppercase()
        return CurrencyInfo(
            code = code,
            name = CurrencyMeta.nameRes[code]?.let { stringProvider.get(it) } ?: code,
            flag = CurrencyMeta.flags[code] ?: "🏳️",
            isCrypto = currency.isCrypto,
        )
    }
}

File: iosApp/iosApp/choose-currency/screen/ChooseCurrencyMapper.swift
import Foundation
import Shared

protocol ChooseCurrencyMapper {
    func toContent(rates: CurrencyRates, recentCodes: [String]) -> ChooseCurrencyContent
}

final class ChooseCurrencyMapperImpl: ChooseCurrencyMapper {
    func toContent(rates: CurrencyRates, recentCodes: [String]) -> ChooseCurrencyContent {
        var allCurrencies: [CurrencyInfo] = []
        allCurrencies.append(toCurrencyInfo(currency: rates.baseCurrency))
        for rate in rates.rates {
            allCurrencies.append(toCurrencyInfo(currency: rate.currency))
        }

        let byCode = Dictionary(allCurrencies.map { ($0.code, $0) }, uniquingKeysWith: { first, _ in first })
        let unique = byCode.values.sorted { $0.code < $1.code }
        let recent = recentCodes.compactMap { byCode[$0] }

        return ChooseCurrencyContent(recentCurrencies: recent, currencies: unique)
    }

    private func toCurrencyInfo(currency: Currency) -> CurrencyInfo {
        let code = currency.name.uppercased()
        return CurrencyInfo(
            code: code,
            name: currencyName(code: code),
            flag: currencyFlag(code: code, isCrypto: currency.isCrypto),
            isCrypto: currency.isCrypto
        )
    }

    private func currencyName(code: String) -> String {
        CurrencyMeta.name(for: code)
    }

    private func currencyFlag(code: String, isCrypto: Bool) -> String {
        CurrencyMeta.flag(for: code, isCrypto: isCrypto)
    }
}
